sgkit/distance/api.py

```diff
@@ -0,0 +1,102 @@
+import dask.array as da
+import numpy as np
+
+from sgkit.distance import metrics
+from sgkit.typing import ArrayLike
+
+
+def pairwise_distance(
+    x: ArrayLike,
+    metric: str = "euclidean",
+) -> np.ndarray:
+    """Calculates the pairwise distance between all pairs of vectors in the
+    given two dimensional array x.
+
+    To illustrate the algorithm consider the following (4, 5) two dimensional array:
+
+    [e.00, e.01, e.02, e.03, e.04]
+    [e.10, e.11, e.12, e.13, e.14]
+    [e.20, e.21, e.22, e.23, e.24]
+    [e.30, e.31, e.32, e.33, e.34]
+
+    The rows of the above matrix are the set of vectors. Now let's label all
+    the vectors as v0, v1, v2, v3.
+
+    The result will be a two dimensional symmetric matrix which will contain
+    the distance between all pairs. Since there are 4 vectors, calculating the
+    distance between each vector and every other vector, will result in 16
+    distances and the resultant array will be of size (4, 4) as follows:
+
+    [v0.v0, v0.v1, v0.v2, v0.v3]
+    [v1.v0, v1.v1, v1.v2, v1.v3]
+    [v2.v0, v2.v1, v2.v2, v2.v3]
+    [v3.v0, v3.v1, v3.v2, v3.v3]
+
+    The (i, j) position in the resulting array (matrix) denotes the distance
+    between vi and vj vectors.
+
+    Negative and nan values are considered as missing values. They are ignored
+    for all distance metric calculations.
+
+    Parameters
+    ----------
+    x
+        [array-like, shape: (M, N)]
+        An array like two dimensional matrix. The rows are the
+        vectors used for comparison, i.e. for pairwise distance.
+    metric
+        The distance metric to use. The distance function can be
+        'euclidean' or 'correlation'
+
+    Returns
+    -------
+
+    [array-like, shape: (M, N)]
+    A two dimensional distance matrix, which will be symmetric. The dimension
+    will be (M, N). The (i, j) position in the resulting array
+    (matrix) denotes the distance between ith and jth vectors.
+
+    Examples
+    --------
+
+    >>> from sgkit.distance.api import pairwise_distance
+    >>> import dask.array as da
+    >>> x = da.array([[6, 4, 1,], [4, 5, 2], [9, 7, 3]]).rechunk(2, 2)
+    >>> pairwise_distance(x, metric='euclidean')
+    array([[0.        , 2.44948974, 4.69041576],
+           [2.44948974, 0.        , 5.47722558],
+           [4.69041576, 5.47722558, 0.        ]])
+
+    >>> import numpy as np
+    >>> x = np.array([[6, 4, 1,], [4, 5, 2], [9, 7, 3]])
+    >>> pairwise_distance(x, metric='euclidean')
+    array([[0.        , 2.44948974, 4.69041576],
+           [2.44948974, 0.        , 5.47722558],
+           [4.69041576, 5.47722558, 0.        ]])
+
+    >>> x = np.array([[6, 4, 1,], [4, 5, 2], [9, 7, 3]])
+    >>> pairwise_distance(x, metric='correlation')
```

Comment:
It is odd for us to call this a distance function and then not invert correlation in some way.  It is a similarity measure here rather than a distance measure, so we should probably be subtracting it from 1 (e.g. https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#Pearson's_distance).

Reply:
Makes sense, I noticed this too, while comparing it with scipy's `pdist(x, metric="correlation")`, then it completely got off my head.